src/core/score_map.cc

```diff
@@ -146,4 +146,43 @@ detail::SdsScorePair ScoreMap::iterator::BreakToPair(void* obj) {
   return detail::SdsScorePair(f, GetValue(f));
 }
 
+namespace {
+// Does not Release obj. Callers must do so explicitly if a `Reallocation` happened
+pair<sds, bool> ReallocIfNeededGeneric(void* obj, float ratio) {
+  sds key = (sds)obj;
+  size_t key_len = sdslen(key);
+
+  if (!zmalloc_page_is_underutilized(key, ratio))
+    return {key, false};
+
+  sds newkey = AllocSdsWithSpace(key_len, 8);
+  memcpy(newkey, key, key_len + 8 + 1);
+
+  return {newkey, true};
+}
+
+}  // namespace
+
+bool ScoreMap::iterator::ReallocIfNeeded(float ratio, std::function<void(sds, sds)> cb) {
+  // Unwrap all links to correctly call SetObject()
+  auto* ptr = curr_entry_;
+
+  if (ptr->IsLink()) {
+    ptr = ptr->AsLink();
+  }
+
+  // Note: we do not iterate over the links. Although we could that...
```

Comment:
yeah, I think we should do that. ptr should be casted into DenseLinkKey for that. Same with sorted_map

Reply:
I agree, I will add a separate PR for this that fixes this for both flows

Reply:
https://github.com/dragonflydb/dragonfly/issues/3949